cmd/download.go

```diff
@@ -0,0 +1,250 @@
+}
+
+// createOutputFileOrDie creates the output file at the given filepath if it does not already exist
+// and returns the file handle.
+// This is a non-destructive operation. Hence, if a file already exists at the given filepath then
+// the command exits with a non-success error code. If any other error case the command exits with
+// a non-success error code as well.
+//
+// Note: The callee has to make sure that the file handle is closed properly.
+func createOutputFileOrDie(filepath string) *os.File {
+	outputFile, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+	if err != nil {
+		if os.IsExist(err) {
+			fmt.Printf("The output file at `%v` does already exist.\n", filepath)
+			os.Exit(3)
+		} else {
+			fmt.Printf("could not open/create the output file at `%s`: %v", filepath, err)
+			os.Exit(4)
+		}
+	}
+	return outputFile
+}
+
+// writeResources writes all resources of the given resourceBundle as-is to the given sink.
+// Returns the number of written resources or an error if the resources can not be written to the sink.
+// An error is only possible if there is an actual issue writing to the file or the given resource
+// bundle is invalid with regards to the FHIR specification.
+func writeResources(resourceBundle map[string]interface{}, sink io.Writer) (resources int, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("encountered a malformed resource bundle")
+		}
+	}()
+
+	entries := resourceBundle["entry"].([]interface{})
+	for e := range entries {
+		entry := entries[e].(map[string]interface{})
+
+		if entry["resource"] != nil {
+			resources++
+			jsonEncoded, err := json.Marshal(entry["resource"])
+			if err != nil {
+				return -1, fmt.Errorf("could not encode resource as JSON: %v\n", err)
+			}
+
+			if _, err = sink.Write(append(jsonEncoded, '\n')); err != nil {
+				return -1, fmt.Errorf("could not write resource to file: %v\n", err)
+			}
+		}
+	}
+
+	return resources, err
+}
+
+// getNextPageURL tries to extract the URL to the next resource bundle page from
+// a given generic resourceBundle as received from a FHIR server.
+// The extraction respects the FHIR specification with regard to how links are
+// defined: https://www.iana.org/assignments/link-relations/link-relations.xhtml#link-relations-1
+//
+// Returns an URL if there is any or nil. Returns an error if the URL can not be extracted.
+// An error is only possible if the pagination URL is malformed.
+func getNextPageURL(resourceBundle map[string]interface{}) (nextURL *url.URL, err error) {
+	if resourceBundle["link"] == nil {
+		return nil, nil
+	}
+
+	links := resourceBundle["link"].([]interface{})
+	for l := range links {
+		linkEntry := links[l].(map[string]interface{})
+
+		if linkEntry["relation"] == "next" {
+			nextPageURL, err := url.ParseRequestURI(linkEntry["url"].(string))
+			if err != nil {
+				return nil, err
+			} else {
+				return nextPageURL, nil
+			}
+		}
+	}
+
+	return nil, err
```

Comment:
is err nil here? above you return nil, nil. is this different?

Reply:
Yes, this is nil. Was a leftover from when this function had a recover block as well. Will be adjusted and the return parameters won't be named any longer.